enhancements/scheduling/prevent-user-workloads-from-being-scheduled-on-control-plane-nodes.md

```diff
@@ -0,0 +1,504 @@
+- User/Developer: Those deploying workloads within the cluster must ensure their applications carry the correct tolerations as advised by Cluster Administrators, especially when targeting special node groups. They need to stay informed about the cluster's scheduling policies and adapt their workloads accordingly.
+
+Additional Considerations:
+
+- Security Auditors: Although not directly involved in the workflow, security auditors need to periodically review and audit the applied ValidatingAdmissionPolicies, ValidatingAdmissionPolicyBindings, and relevant RBAC rules to ensure compliance with the intended security and operational policies. They may also review logs and events related to admission policy rejections to ensure policies are correctly enforced.
+
+#### Pre-Workflow Steps
+
+1. Continuous Standardization of Shipped tolerations.
+
+- OpenShift Engineers need to ensure that all control plane and essential workloads ship with the correct tolerations for the NoExecute taints being applied to control plane nodes.
+This is critical for maintaining uninterrupted operations of OpenShift's core services on these nodes.
+This involves getting multiple teams on the same page and that new projects also incorporate the needed tolerations.
+- OpenShift Engineers also need to guarantee that those workloads have the right permissions to tolerate the NoExecute taint, as the ValidatingAdmissionPolicy checks for custom taint/toleration key and verbs.
+- If custom special node groups are being considered for protection, Cluster Administrators need to coordinate with Users/Developers to implement new tolerations for workloads intended for those specific node groups (e.g., GPU-enabled nodes).
+This coordination ensures that the designated workloads are appropriately scheduled on the protected nodes.
+
+#### Workflow Steps
+
+1. Node tainting:
+
+- Cluster Administrator applies NoExecute taints to control plane nodes and any other special node groups deemed necessary to protect.
+This foundational step prevents unauthorized workloads from being scheduled (or directly placed with spec.nodeName field) on these critical nodes.
+
+```
+oc taint nodes ip-XX-XX-XX-XXX.ec2.internal node-role.kubernetes.io/control-plane:NoExecute-
+```
+
+2. Validating Admission Policy provided:
+
+- As part of this proposal a `ValidatingAdmissionPolicy` will be provided as part of the static manifests applied by Kube Scheduler Operator ([staticresourcecontroller](https://github.com/openshift/cluster-kube-scheduler-operator/blob/67304344622e50cbcd95e7a05056294e9d96df99/pkg/operator/starter.go#L108)).
+
+```
+apiVersion: admissionregistration.k8s.io/v1beta1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "control-plane-scheduling-policy"
+spec:
+  failurePolicy: Fail
+  matchConstraints:
+    resourceRules:
+    - apiGroups:   [""]
+      apiVersions: ["v1"]
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["pods"]
+  validations:
```

Comment:
What about the mirror pods?
- if the manifest is on the node, they will run regardless, so there is no point in validating the permissions here
- static pods do not have service accounts assigned

We could identify that the request originated from the kubelet and only allow it to create the mirror pods? 

Mirror pods have a `kubernetes.io/config.mirror` annotation.  Foreign pods with this annotation are deleted by the kubelet. If we want to be more strict on admission we could also enable [the node restriction plugin](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#noderestriction).

Reply:
I think it would be safer to check for the kubelet SA as the plugin does to prevent the creation of mirror pods by other actors other than a kubelet. As it would limit the surface for security bugs.

That is why I mentioned the node-restriction-plugin as it limits who can create these annotations. It is not a required piece to have, but it would get us halfway there.

Reply:
So you are thinking about something like this? 
```
(request.userInfo.username == 'system:serviceaccount:kube-system:kubelet' || 
         request.userInfo.username == 'system:kubelet') ||
```
I tested it and it seems to work.

While the node-restriction plugin enhances security by restricting who can set certain annotations related to node or kubelet functionalities, mandating its usage might not be flexible for all customers, right? We can include this as a recommended security practice in the documentation, also here in the KEP, especially for environments where static pods are commonly used or where enhanced security controls are necessary, wdyt? This approach provides guidance without enforcing a specific operational model, allowing customers to choose based on their security posture and operational requirements. I know the approach proposed here is already being designed with security posture in mind, but this all or nothing for he plugin enablement on top of the other stuff sounds weird to me. They already have means to enable it right?

@deads2k @atiratree what do you think?

Reply:
While we want to avoid blocking API updates for static pods, I wouldn't over rotate on it.  An exclusion for those users in the `system:nodes` group when mutating mirror pods seems doable, but if it ends up overly complex then living with that drift between reality and the API is acceptable, even if unfortunate.

Reply:
I misread the node restriction plugin and it does not affect pods not originating from the node. Also it is enabled by default: https://github.com/openshift/cluster-kube-apiserver-operator/blob/58d79ab9d65972bbd706e0787d149e7c0aef1774/bindata/assets/config/defaultconfig.yaml#L65

+1, it seems that doing the same check as noderestriction plugin does should not be a big hassle: https://github.com/kubernetes/kubernetes/blob/master/pkg/auth/nodeidentifier/default.go

Reply:
Ok, so I think I could do something like 
`request.userInfo.groups.contains('system:nodes') ||`

@benluddy or is there any better way to do it?

Reply:
Yes, you can use the `in` operator (`contains` is a method on `string`), like this: `"system:nodes" in request.userInfo.groups`.

Reply:
Thanks, that makes sense :) I guess I would have figured that out when I ran it :sweat_smile: 

I was also asking about the general approach of this thread